### PR TITLE
Implement Phase 2 Muon HLT validation

### PR DESCRIPTION
**Pull request by Parsifal-2045**
#### PR description:

This PR implements the full Muon HLT Validation for Phase 2.
When a Phase 2 workflow is run, the validation module's output changes from Phase 1 objects to the corresponding Phase 2 ones. The HLT full event content has been updated accordingly and a few more plots have been added (e.g separate efficiency measurements in the barrel, overlap, and endcap).
Changes presented in the  [HLT Upgrade meeting on 3/12/2024](https://indico.cern.ch/event/1482695/contributions/6247483/subcontributions/517161/attachments/2978374/5243554/2024-12-03%20Phase%202%20validation%20and%20integration%20in%20CMSSW.pdf).

#### PR validation:

Verified that validation plots are filled with sensible information using ZMM and TTbar RelVals and private production via runTheMatrix using workflows 29834 (TTbar) and 29850 (ZMM) at 200 PU.

### CODE REVIEW

File: HLTrigger/Configuration/python/HLTrigger_EventContent_cff.py [hunk 1]
```diff
@@ -400,7 +400,15 @@
         'keep *_hltScoutingPFPacker_*_*',
         'keep *_hltScoutingPrimaryVertexPacker_*_*',
         'keep *_hltScoutingTrackPacker_*_*',
-        'keep edmTriggerResults_*_*_*'
+        'keep edmTriggerResults_*_*_*',
+        'keep *_hltL2MuonSeedsFromL1TkMuon_*_*',
```
Comment (mmusich) [inline HLTrigger/Configuration/python/HLTrigger_EventContent_cff.py:404]: please do NOT change this file directly as it is used also for the Run3 HLT menu and its content synchronized with confDB. Change instead `Configuration/EventContent/python/EventContent_cff.py` as done e.g. in https://github.com/cms-sw/cmssw/pull/46823

**Comment by mmusich** [comment]
-hlt
   * see https://github.com/cms-sw/cmssw/pull/46860#discussion_r1869060089

File: HLTriggerOffline/Common/python/HLTValidationHarvest_cff.py [hunk 2]
```diff
@@ -42,11 +42,11 @@
 # Temporary Phase-2 configuration
 # Exclude everything except JetMET for now
 from Configuration.Eras.Modifier_phase2_common_cff import phase2_common
-phase2_common.toReplaceWith(hltpostvalidation, hltpostvalidation.copyAndExclude([HLTMuonPostVal,
+phase2_common.toReplaceWith(hltpostvalidation, hltpostvalidation.copyAndExclude([#HLTMuonPostVal,
```
Comment (mmusich) [inline HLTriggerOffline/Common/python/HLTValidationHarvest_cff.py:45]: why keeping it commented?

File: HLTriggerOffline/Common/python/HLTValidationHarvest_cff.py [hunk 3]
```diff
@@ -42,11 +42,11 @@
...
                                                                                  HLTTauPostVal,
                                                                                  EgammaPostVal,
                                                                                  postProcessorHLTgsfTrackingSequence,
-                                                                                 postProcessorHLTmuonTrackingSequence,
+                                                                                 #postProcessorHLTmuonTrackingSequence,
```
Comment (mmusich) [inline HLTriggerOffline/Common/python/HLTValidationHarvest_cff.py:49]: why keeping it commented?

File: Validation/RecoMuon/python/associators_cff.py [hunk 4]
```diff
@@ -174,8 +174,8 @@
 #    acceptOneStubMatchings = cms.bool(False),
 #    rejectBadGlobal = cms.bool(True),
 ##############################################
-    PurityCut_track = 0.75,
-    PurityCut_muon = 0.75,
+    PurityCut_track = cms.double(0.75),
+    PurityCut_muon = cms.double(0.75),
```
Comment (mmusich) [inline Validation/RecoMuon/python/associators_cff.py:178]: why is this needed?

File: Validation/RecoMuon/python/PostProcessorHLT_cff.py [hunk 5]
```diff
@@ -4,25 +4,32 @@
 from DQMServices.Core.DQMEDHarvester import DQMEDHarvester
 
 from Validation.RecoMuon.PostProcessor_cff import postProcessorMuonTrack
+
 postProcessorMuonTrackHLT = postProcessorMuonTrack.clone(
-    subDirs = ["HLT/Muon/MuonTrack/*"]
+    subDirs=["HLT/Muon/MuonTrack/*"]
```
Comment (mmusich) [inline Validation/RecoMuon/python/PostProcessorHLT_cff.py:9]: why are the spaces around `=` signs removed? There's no universal consensus on the python formatting (see https://github.com/cms-sw/cmssw/issues/45706) but I think this option is clearly disfavored.

File: Validation/RecoMuon/plugins/MuonTrackValidator.cc [hunk 6]
```diff
@@ -108,6 +160,20 @@ void MuonTrackValidator::bookHistograms(DQMEDAnalyzer::DQMStore::IBooker& ibooke
                                           minNHit,
                                           maxNHit));
 
+      int bindR{200};
+      double mindR{0.0};
+      double maxdR{10.0};
+      ;
```
Comment (mmusich) [inline Validation/RecoMuon/plugins/MuonTrackValidator.cc:166]:
```suggestion
```

File: HLTrigger/Configuration/python/HLTrigger_EventContent_cff.py [hunk 7]
```diff
@@ -400,7 +400,15 @@
...
+        'keep *_hltL2MuonsFromL1TkMuon_*_*',
+        'keep *_hltIter2Phase2L3FromL1TkMuonMerged_*_*',
+        'keep *_hltPhase2L3OIMuonTrackSelectionHighPurity_*_*',
+        'keep *_hltPhase2L3MuonMerged_*_*',
+        'keep *_hltPhase2L3GlbMuon_*_*',
+        'keep *_hltPhase2L3MuonsNoID_*_*',
+        'keep *_hltPhase2L3Muons_*_*'
```
Comment (Martin-Grunewald) [inline HLTrigger/Configuration/python/HLTrigger_EventContent_cff.py:411]: This is an auto-generated file for Run3, any changes will be lost! Please remove the change to this file.

Instead, you'd need to add your changes, using an era/modifier approach, here:
https://github.com/cms-sw/cmssw/blob/master/Configuration/EventContent/python/EventContent_cff.py
see there for examples (eg, Phase2 tracker etc.).

File: Validation/RecoMuon/plugins/MuonTrackValidator.cc [hunk 8]
```diff
@@ -475,6 +541,16 @@ void MuonTrackValidator::analyze(const edm::Event& event, const edm::EventSetup&
       int ats = 0;
       int st = 0;
 
+      std::optional<double> dR = [&]() -> std::optional<double> {
+        if (tPC.size() == 2) {
+          TrackingParticle::Vector p1 = lhcParametersDefinerTP_->momentum(event, setup, tPC[0]);
+          TrackingParticle::Vector p2 = lhcParametersDefinerTP_->momentum(event, setup, tPC[1]);
+
+          return reco::deltaR(p1, p2);
```
Comment (mmusich) [inline Validation/RecoMuon/plugins/MuonTrackValidator.cc:549]: using `deltaR2` is computationally convenient.

**Review by mmusich — COMMENTED** [review]
few other comments.

File: HLTrigger/Configuration/python/HLTrigger_EventContent_cff.py (hunk 7 above)
Comment (mmusich) [reply HLTrigger/Configuration/python/HLTrigger_EventContent_cff.py:411]: @Martin-Grunewald already commented at https://github.com/cms-sw/cmssw/pull/46860#discussion_r1869060089

File: HLTriggerOffline/Common/python/HLTValidationHarvest_cff.py (hunk 3 above)
Comment (Parsifal-2045) [reply HLTriggerOffline/Common/python/HLTValidationHarvest_cff.py:49]: In both cases, I've followed the same approach as the other validation sequences (that I assumed were commented each time one got integrated). There would be no problem outright removing them

File: Validation/RecoMuon/python/associators_cff.py (hunk 4 above)
Comment (Parsifal-2045) [reply Validation/RecoMuon/python/associators_cff.py:178]: No reason in particular. The defaults were given in cms.double() form so I did the same

File: HLTriggerOffline/Common/python/HLTValidationHarvest_cff.py (hunk 3 above)
Comment (mmusich) [reply HLTriggerOffline/Common/python/HLTValidationHarvest_cff.py:49]: I'd recommend to remove them, given this PR generally addresses adding back the muon object DQM/ validation sequences (and also to remove unncessary clutter).

**Comment by mmusich** [comment]
@cmsbuild please test

**Comment by mmusich** [comment]
@Parsifal-2045 thank you for the prompt updates.

**Comment by mmusich** [comment]
Looks like fastsim cannot digest the changes in `SeedToTrackProducer`
See e.g.:  [here](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-76346f/43237/addOnTests/runall-report.log) and [here](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-76346f/43237/addOnTests/fastsim1/step1.log).

**Comment by Parsifal-2045** [comment]
> Looks like fastsim cannot digest the changes in `SeedToTrackProducer` See e.g.: [here](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-76346f/43237/addOnTests/runall-report.log) and [here](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-76346f/43237/addOnTests/fastsim1/step1.log).

Looking into it. It seems the `ancientMuonSeed` collection is not always produced for some reason

**Comment by Parsifal-2045** [comment]
The issue should now be solved, with two different implementations of the `SeedToTrackProducer` for Phase 1 and Phase 2

**Comment by Parsifal-2045** [comment]
> > The issue should now be solved, with two different implementations of the `SeedToTrackProducer` for Phase 1 and Phase 2
> 
> I am not a big fan of code duplication. I haven't looked yet at the full implementation but if it's just because of the signature of the methods, can't it be a templated code or some kind of traits class?

I can have a look at that

**Comment by Parsifal-2045** [comment]
This commit implements the requested templated version of the `SeedToTrackProducer` class. The base class contains the full implementation, while `SeedToTrackProducer` and `Phase2SeedToTrackProducer` are specialized for Phase 1 and Phase 2 seeds, respectively

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
> -1

sadly, failures are unrelated.

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by antoniovagnerini** [comment]
+1

**Comment by civanch** [comment]
+1

i also not a fun of duplications but we already have such cases for Phase1 and Phase2 in other corners of CMSSW with the advantage of independence of the production Phase1 code from Phase2 developments.

**Comment by mandrenguyen** [comment]
+1

**Comment by iarspider** [comment]
@Parsifal-2045 this PR causes linking issues in CLANG_X IBs: https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/el8_amd64_gcc12/CMSSW_15_0_CLANG_X_2024-12-08-2300/SimMuon/MCTruth

**Comment by mmusich** [comment]
```diff
diff --git a/SimMuon/MCTruth/plugins/Phase2SeedToTrackProducer.h b/SimMuon/MCTruth/plugins/Phase2SeedToTrackProducer.h
index 12fe7141af5..9e223a9d33c 100644
--- a/SimMuon/MCTruth/plugins/Phase2SeedToTrackProducer.h
+++ b/SimMuon/MCTruth/plugins/Phase2SeedToTrackProducer.h
@@ -11,7 +11,10 @@
 
 #include "SimMuon/MCTruth/plugins/SeedToTrackProducerBase.h"
 #include "DataFormats/MuonSeed/interface/L2MuonTrajectorySeedCollection.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/PluginManager/interface/ModuleDef.h"
 
 typedef SeedToTrackProducerBase<L2MuonTrajectorySeedCollection> Phase2SeedToTrackProducer;
 
+DEFINE_FWK_MODULE(Phase2SeedToTrackProducer);
 #endif
diff --git a/SimMuon/MCTruth/plugins/SealModule.cc b/SimMuon/MCTruth/plugins/SealModule.cc
index b2988b656c8..cf557981fe9 100644
--- a/SimMuon/MCTruth/plugins/SealModule.cc
+++ b/SimMuon/MCTruth/plugins/SealModule.cc
@@ -2,10 +2,6 @@
 #include "FWCore/PluginManager/interface/ModuleDef.h"
 #include "SimMuon/MCTruth/interface/CSCTruthTest.h"
 #include "SimMuon/MCTruth/plugins/MuonAssociatorEDProducer.h"
-#include "SimMuon/MCTruth/plugins/SeedToTrackProducer.h"
-#include "SimMuon/MCTruth/plugins/Phase2SeedToTrackProducer.h"
 
 DEFINE_FWK_MODULE(MuonAssociatorEDProducer);
 DEFINE_FWK_MODULE(CSCTruthTest);
-DEFINE_FWK_MODULE(SeedToTrackProducer);
-DEFINE_FWK_MODULE(Phase2SeedToTrackProducer);
diff --git a/SimMuon/MCTruth/plugins/SeedToTrackProducer.h b/SimMuon/MCTruth/plugins/SeedToTrackProducer.h
index c70eb1f14e8..5fc4db30b04 100644
--- a/SimMuon/MCTruth/plugins/SeedToTrackProducer.h
+++ b/SimMuon/MCTruth/plugins/SeedToTrackProducer.h
@@ -11,7 +11,11 @@
 
 #include "SimMuon/MCTruth/plugins/SeedToTrackProducerBase.h"
 #include "DataFormats/TrajectorySeed/interface/TrajectorySeedCollection.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/PluginManager/interface/ModuleDef.h"
 
 typedef SeedToTrackProducerBase<TrajectorySeedCollection> SeedToTrackProducer;
 
+DEFINE_FWK_MODULE(SeedToTrackProducer);
+
 #endif
```

solves, by the way I think we should get rid of the `SealModule.cc` entirely.

**Comment by iarspider** [comment]
@mmusich could you please open a PR with this fix? Thanks!

**Comment by mmusich** [comment]
https://github.com/cms-sw/cmssw/pull/46902

**Comment by dan131riley** [comment]
We're also getting segfaults in ROOT6 wf 24034.0 that may be related.

CMSSW_15_0_ROOT6_X_2024-12-08-2300:

```#2  0x000014803c0c1704 in sig_dostack_then_abort () from /cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc12/cms/cmssw/CMSSW_15_0_ROOT6_X_2024-12-08-2300/lib/el8_amd64_gcc12/pluginFWCoreServicesPlugins.so
#3  <signal handler called>
#4  0x0000147fc9938a85 in SeedToTrackProducerBase<std::vector<L2MuonTrajectorySeed, std::allocator<L2MuonTrajectorySeed> > >::produce(edm::StreamID, edm::Event&, edm::EventSetup const&) const () from /cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc12/cms/cmssw/CMSSW_15_0_ROOT6_X_2024-12-08-2300/lib/el8_amd64_gcc12/pluginSimMuonMCTruthPlugins.so
```
Preceded by
```
%MSG-e SeedToTrackProducerBase:   Phase2SeedToTrackProducer:hltPhase2L2MuonSeedTracks 09-Dec-2024 07:54:05 CET  Run: 1 Event: 1
L2 seeds collection not found !! 
```

**Comment by mmusich** [comment]
> We're also getting segfaults in ROOT6 wf 24034.0 that may be related.

isn't it already addressed at https://github.com/cms-sw/cmssw/pull/46896 ?

**Comment by dan131riley** [comment]
> > We're also getting segfaults in ROOT6 wf 24034.0 that may be related.
> 
> isn't it already addressed at #46896 ?

Yes, looks like it.  Somehow I missed that PR while checking if it had already been addressed.
